proxy/blocks.go

```diff
@@ -1,117 +1,113 @@
 
 import (
+	"errors"
 	"log"
 	"math/big"
-	"strconv"
-	"strings"
+
 	"sync"
 
-	"github.com/ethereum/go-ethereum/common"
+	"github.com/dominant-strategies/go-quai/common"
+	"github.com/dominant-strategies/go-quai/core/types"
 
-	"github.com/J-A-M-P-S/go-etcstratum/rpc"
-	"github.com/J-A-M-P-S/go-etcstratum/util"
+	"github.com/dominant-strategies/go-quai-stratum/rpc"
+	"github.com/dominant-strategies/go-quai/common/hexutil"
 )
 
 const maxBacklog = 3
 
 type heightDiffPair struct {
 	diff   *big.Int
-	height uint64
+	height []uint64
 }
 
 type BlockTemplate struct {
 	sync.RWMutex
-	Header               string
-	Seed                 string
-	Target               string
+	Header               *types.Header
+	Target               *big.Int
 	Difficulty           *big.Int
-	Height               uint64
-	GetPendingBlockCache *rpc.GetBlockReplyPart
+	Height               []*big.Int
+	GetPendingBlockCache *rpc.GetBlockReply
 	nonces               map[string]bool
 	headers              map[string]heightDiffPair
 }
 
 type Block struct {
-	difficulty  *big.Int
+	difficulty  []*hexutil.Big
 	hashNoNonce common.Hash
 	nonce       uint64
-	mixDigest   common.Hash
 	number      uint64
 }
 
-func (b Block) Difficulty() *big.Int     { return b.difficulty }
-func (b Block) HashNoNonce() common.Hash { return b.hashNoNonce }
-func (b Block) Nonce() uint64            { return b.nonce }
-func (b Block) MixDigest() common.Hash   { return b.mixDigest }
-func (b Block) NumberU64() uint64        { return b.number }
+func (b Block) Difficulty() []*hexutil.Big { return b.difficulty }
+func (b Block) HashNoNonce() common.Hash   { return b.hashNoNonce }
+func (b Block) Nonce() uint64              { return b.nonce }
+func (b Block) NumberU64() uint64          { return b.number }
 
 func (s *ProxyServer) fetchBlockTemplate() {
-	rpc := s.rpc()
+	rpc := s.rpc(common.ZONE_CTX)
 	t := s.currentBlockTemplate()
-	pendingReply, height, diff, err := s.fetchPendingBlock()
+	pendingHeader, err := rpc.GetWork()
 	if err != nil {
-		log.Printf("Error while refreshing pending block on %s: %s", rpc.Name, err)
-		return
-	}
-	reply, err := rpc.GetWork()
-	if err != nil {
-		log.Printf("Error while refreshing block template on %s: %s", rpc.Name, err)
+		log.Printf("Error while getting pending header (work) on %s: %s", rpc.Name, err)
 		return
 	}
 	// No need to update, we have fresh job
-	if t != nil && t.Header == reply[0] {
+	if t != nil && t.Header == pendingHeader {
 		return
+	} else if t != nil {
+		t.Header = pendingHeader
 	}
 
-	pendingReply.Difficulty = util.ToHex(s.config.Proxy.Difficulty)
-
 	newTemplate := BlockTemplate{
-		Header:               reply[0],
-		Seed:                 reply[1],
-		Target:               reply[2],
-		Height:               height,
-		Difficulty:           big.NewInt(diff),
-		GetPendingBlockCache: pendingReply,
-		headers:              make(map[string]heightDiffPair),
+		Header:     pendingHeader,
+		Target:     pendingHeader.DifficultyArray()[2],
```

Comment:
Please use the Helper functions. Header.Difficulty(2), use Header.Difficulty(common.ZONE_CTX)